demo/llama/train.py

```diff
@@ -0,0 +1,389 @@
+
+# dataflux vars
+project_name = "<YOUR-PROJECT>"
+bucket_name = "<YOUR-BUCKET>"
+checkpoint_save_dir = "<YOUR-BUCKET>"
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+save_interval = 1000
+eval_interval = 1000
+eval_iters = 100
+log_interval = 1
+
+# compile = False
+
+# Hyperparameters
+learning_rate = 6e-4
+batch_size = 125
+micro_batch_size = 5
+max_iters = 600000  # num_epochs * (epoch_size // micro_batch_size) // devices
+weight_decay = 1e-1
+beta1 = 0.9
+beta2 = 0.95
+grad_clip = 1.0
+decay_lr = True
+warmup_iters = 2000
+lr_decay_iters = max_iters
+min_lr = 6e-5
+
+# Data proportions from https://arxiv.org/pdf/2302.13971.pdf Table 1
+data_config = [
+    ("arxiv", 2.5),
+    ("book", 4.5),
+    ("c4", 15.0),
+    ("cc", 67.0),
+    ("github", 4.5),
+    ("stackexchange", 2.0),
+    ("wikipedia", 4.5),
+]
+
+
+def main(
+    devices: int = 4,
+    train_data_dir: Path = "data/lit-redpajama",
+    val_data_dir: Optional[Path] = None,
+) -> None:
+    auto_wrap_policy = partial(transformer_auto_wrap_policy,
+                               transformer_layer_cls={Block})
+    strategy = DatafluxFSDPStrategy(
+        project_name=project_name,
+        storage_client=storage.Client(project=project_name),
+        auto_wrap_policy=auto_wrap_policy,
+        activation_checkpointing=Block,
+        limit_all_gathers=True,
+    )
+
+    fabric = L.Fabric(accelerator="cuda",
+                      devices=devices,
+                      precision="bf16-mixed",
+                      strategy=strategy)
+    fabric.launch()
+    fabric.seed_everything(1337)
+
+    config = LLaMAConfig.from_name("7B")
+
+    train_dataloader, val_dataloader = create_dataloaders(
+        batch_size=micro_batch_size,
+        block_size=config.block_size,
+        fabric=fabric,
+        train_data_dir=train_data_dir,
+        val_data_dir=val_data_dir,
+        seed=1338,
+    )
+    if val_dataloader is None:
+        train_dataloader = fabric.setup_dataloaders(train_dataloader)
+    else:
+        train_dataloader, val_dataloader = fabric.setup_dataloaders(
+            train_dataloader, val_dataloader)
+
+    with fabric.device:
+        torch.set_default_dtype(torch.bfloat16)
+        model = LLaMA(config)
+        model.apply(model._init_weights)
+        torch.set_default_dtype(torch.float32)
+
+    # if compile:
+    #     model = torch.compile(model)
+
+    optimizer = torch.optim.AdamW(
+        model.parameters(),
+        lr=learning_rate,
+        weight_decay=weight_decay,
+        betas=(beta1, beta2),
+        foreach=False,
+    )
+
+    model, optimizer = fabric.setup(model, optimizer)
+
+    process_batch_size = batch_size // devices
+    gradient_accumulation_iters = process_batch_size // micro_batch_size
+
+    train(fabric, model, optimizer, train_dataloader, val_dataloader,
+          gradient_accumulation_iters, devices)
+
+
+def train(
```

Comment:
Can the underlying implementation be reused here? It looks like there might not be any changes